src/actions/mod.rs

```diff
@@ -211,6 +213,51 @@ fn find_word_at_pos(line: &str, pos: &Column) -> (Column, Column) {
     (span::Column::new_zero_indexed(start), span::Column::new_zero_indexed(end))
 }
 
+// TODO include workspace Cargo.tomls in watchers / relevant
+/// Client file-watching request / filtering logic
+/// We want to watch workspace 'Cargo.toml', root 'Cargo.lock' & the root 'target' dir
+pub struct FileWatch<'ctx> {
+    project_str: &'ctx str,
+    project_uri: String,
+}
+
+impl<'ctx> FileWatch<'ctx> {
+    pub fn new(ctx: &'ctx InitActionContext) -> Self {
+        Self {
+            project_str: ctx.current_project.to_str().unwrap(),
+            project_uri: Url::from_file_path(&ctx.current_project).unwrap().into_string(),
+        }
+    }
+
+    /// Returns json config for desired file watches
+    pub fn watchers_config(&self) -> serde_json::Value {
+        let pattern = format!("{}/Cargo{{.toml,.lock}}", self.project_str);
+        let target_pattern = format!("{}/target", self.project_str);
+        // For target, we only watch if it gets deleted.
+        json!({
+            "watchers": [{ "globPattern": pattern }, { "globPattern": target_pattern, "kind": 4 }]
+        })
+    }
+
+    /// Returns if a file change is relevant to the files we actually wanted to watch
+    // Implementation note: This is expected to be called a large number of times in a loop
+    // so should be fast / avoid allocation.
+    #[inline]
+    pub fn is_relevant(&self, change: &FileEvent) -> bool {
+        let path = change.uri.as_str();
+
+        if !path.starts_with(&self.project_uri) {
+            return false;
+        }
+
+        let local = &path[self.project_uri.len()..];
+
+        local == "/Cargo.lock" || local == "/Cargo.toml"
```

Comment:
~~While I understand that changing `Cargo.{lock, toml}` files should warrant a Cargo rebuild, I think it's also reasonable not to ignore any `*.rs` file here, even at risk of being a bit over-eager. I think it's safe to do here and this can happen a couple of times, e.g. when files are changed via source control or when a quick fix might've been applied outside current editor.~~
Note to self: previous eager rebuilds on every change might've triggered unnecessary Cargo rebuilds due to build scripts and similar, so maybe it's important to ignore build script-emitted files.
While it'd be nice to still watch `*.rs` files, I think we can't reliably do that unless we analyze what changed externally and what changed via build, so disregard what I said and I'll just leave it here to have that in mind.